linera-core/src/client/mod.rs

```diff
@@ -1221,91 +1223,124 @@ where
-                            "Postponing received certificate from {:.8} at height {} \
-                            from future epoch {}",
-                            chain_id,
-                            certificate.value().height(),
-                            certificate.value().epoch()
-                        );
-                        // Stop the synchronization here. Do not increment the tracker further so
-                        // that this certificate can still be downloaded later, once our committee
-                        // is updated.
-                    }
-                    HandleCertificateResult::OldEpoch => {
-                        tracker.push(chain_id, certificate.value().height());
-                        // This epoch is not recognized any more. Let's skip the certificate.
-                        // If a higher block with a recognized epoch comes up later from the
-                        // same chain, the call to `receive_certificate` below will download
-                        // the skipped certificate again.
-                        warn!(
-                            "Skipping received certificate from past epoch {:?}",
-                            certificate.value().epoch()
-                        );
-                    }
-                    HandleCertificateResult::New => {
-                        tracker.push(chain_id, certificate.value().height());
-                        certificates.push(certificate);
-                    }
+        let Some(last_height) = block_batch.last().copied() else {
+            return Ok(Vec::new());
+        };
+
+        let batch_size = last_height.saturating_sub(first_height).0 + 1;
+        let block_batch_range = BlockHeightRange::multi(first_height, batch_size);
+        let query = ChainInfoQuery::new(chain_id)
+            .with_sent_certificate_hashes_in_range(block_batch_range.clone());
+
+        let remote_response = remote_node.handle_chain_info_query(query).await?;
+
+        let certificate_hashes = match remote_response.requested_sent_certificate_hashes.as_slice()
+        {
+            hashes if hashes.len() as u64 != block_batch_range.limit.unwrap() => {
+                error!(
+                    block_range = ?block_batch_range,
+                    received_num = hashes.len(),
+                    "Validator sent invalid number of certificate hashes."
+                );
+                return Err(NodeError::InvalidChainInfoResponse);
+            }
+            hashes => hashes.to_vec(),
+        };
+
+        let remote_certificates = remote_node
+            .node
+            .download_certificates(certificate_hashes)
+            .await?;
+
+        let mut certificates = Vec::new();
+
+        for certificate in remote_certificates {
+            match self
+                .check_certificate(max_epoch, committees, &certificate)
+                .await?
+            {
+                HandleCertificateResult::FutureEpoch => {
```

Comment:
We should probably rename this to `CheckCertificateResult`. (The certificate wasn't handled yet, i.e. we haven't executed the block and updated the chain state yet.)